Prioritize UDF over scalar built-in function in case of function name…

epsio-banay opened the pull request:
# Which issue does this PR close?

Closes #6588.

# Rationale for this change
This PR lets a user override the built-in implementation of scalar functions (such as abs, floor, etc) using the ScalarUdf feature.
We are currently only using the logical planner of Datafusion, this feature would allow us and others to have more control over the built-in scalar functions.
It should also be inline with Datafusion's intention to be customizable.

<!--
 Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
 Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.  
-->

# What changes are included in this PR?

Scalar UDFs will have precedence over Scalar built-in functions in case of name conflicts.
For example if a user registers a UDF with the name "abs" (as the built-in function called "abs"), The sql "SELECT abs(10)" will use the UDF instead of the built-in function.

<!--
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->

# Are these changes tested?

Yes
<!--
We typically require tests for all PRs in order to:
1. Prevent the code from being accidentally broken by subsequent changes
2. Serve as another way to document the expected behavior of the code

If tests are not included in your PR, please explain why (for example, are they covered by existing tests)?
-->

# Are there any user-facing changes?

Yes.
DataFusion will prioritize UDF over built-in scalar function in case of function name conflict.

<!--
If there are user-facing changes then we may require documentation to be updated before approving the PR.
-->

<!--
If there are any breaking changes to public APIs, please add the `api change` label.
-->

alamb commented on datafusion/common/src/config.rs:
```diff
@@ -191,6 +191,8 @@ config_namespace! {
         /// MySQL, PostgreSQL, Hive, SQLite, Snowflake, Redshift, MsSQL, ClickHouse, BigQuery, and Ansi.
         pub dialect: String, default = "generic".to_string()
 
+        /// Should DataFusion prioritize UDF over built-in scalar function in case of function name conflict.
```
I actually don't think a config flag is needed -- if a UDF is registered with the same name as a built in, it will not be callable (as you note in the issue / PR description). Therefore I think it is safe to assume that the if a user registers a function with the same name as a built in they want the new function to have precidence

ozankabak replied:
Agreed

epsio-banay replied:
Done.
Thanks for responding so fast, this PR got much more simple.
I'll also edit the PR description - remove the described user facing changes

alamb reviewed (commented):
Thank you for the contribution @epsio-banay ! I left some comments -- let me know if they don't make sense

alamb commented:
Marking as draft as this PR has feedback and is awaiting a response (I am trying to keep the list of PRs needing review clear)

epsio-banay commented on datafusion/sql/tests/integration_test.rs:
```diff
@@ -125,12 +127,78 @@ fn parse_ident_normalization() {
+    ));
+
+    // Test data
+    let test_data = [
```
Done in scalar_udf_override_built_in_scalar_function

alamb reviewed (approved):
Looks great to me @epsio-banay  -- thank you very much for the contribution

alamb commented:
I think this branch has a logical merge conflict with  some of the changes on master. If I applied this diff locally the tests passed:

```diff
diff --git a/datafusion/core/tests/sql/udf.rs b/datafusion/core/tests/sql/udf.rs
index 0ecd5d0fd..eb76c3417 100644
--- a/datafusion/core/tests/sql/udf.rs
+++ b/datafusion/core/tests/sql/udf.rs
@@ -185,7 +185,7 @@ async fn scalar_udf_override_built_in_scalar_function() -> Result<()> {
 
     let batch = RecordBatch::try_new(
         Arc::new(schema.clone()),
-        vec![Arc::new(Int32Array::from(vec![-100]))],
+        vec![Arc::new(Int32Array::from_slice([-100]))],
     )?;
     let ctx = SessionContext::new();
 
```

epsio-banay commented:
> I think this branch has a logical merge conflict with some of the changes on master. If I applied this diff locally the tests passed:
> 
> ```diff
> diff --git a/datafusion/core/tests/sql/udf.rs b/datafusion/core/tests/sql/udf.rs
> index 0ecd5d0fd..eb76c3417 100644
> --- a/datafusion/core/tests/sql/udf.rs
> +++ b/datafusion/core/tests/sql/udf.rs
> @@ -185,7 +185,7 @@ async fn scalar_udf_override_built_in_scalar_function() -> Result<()> {
>  
>      let batch = RecordBatch::try_new(
>          Arc::new(schema.clone()),
> -        vec![Arc::new(Int32Array::from(vec![-100]))],
> +        vec![Arc::new(Int32Array::from_slice([-100]))],
>      )?;
>      let ctx = SessionContext::new();
>  
> ```

rebased and fixed:)

alamb commented:
Thanks again @epsio-banay
